base/channels.jl

```diff
@@ -553,6 +664,127 @@ function take_unbuffered(c::Channel{T}) where T
+If `buffer` is provided, it will be used to store the result. If the channel closes before
+`n` items are taken, the result will contain only the items that were available. For
+buffered channels, this operation requires fewer `lock` operations than individual `take!(c)`s.
+
+# Examples
+
+```jldoctest
+julia> c = Channel(3);
+
+julia> append!(c, 1:3);
+
+julia> take!(c, 3, zeros(Int, 3))
+[1,2,3]
+```
+
+!!! compat "Julia 1.12"
+    Requires at least Julia 1.12.
+"""
+function take!(c::Channel{T}, n::Integer) where {T}
+    return _take(c, n, Vector{T}(undef, n))
+end
+function take!(c::Channel{T}, n::Integer, buffer::AbstractVector{T2}) where {T2,T<:T2}
+    # buffer is user defined, so make sure it has the correct size
+    if length(buffer) != n
+        resize!(buffer, n)
+    end
+    return _take(c, n, buffer)
+end
+
+function _take(c::Channel{T}, n::Integer, buffer::AbstractVector) where {T}
+    buffered = isbuffered(c)
+    # short-circuit for small n
+    if n == 0
+        return buffer
+    elseif n == 1
+        try
+            x = take!(c)
+            @inbounds buffer[begin] = x
+        catch e
+            # when provided a batch size, take should not error on a closed column
+            # so we need to return an empty buffer if we fail to get one element.
+            if e isa InvalidStateException && e.state === :closed
+                empty!(buffer)
+                return buffer
+            else
+                rethrow()
+            end
+        end
+        return buffer
+    end
+    buffered ? take_buffered(c, n, buffer) : take_unbuffered(c, buffer)
+end
+
+function take_buffered(c::Channel{T}, n::Integer, buffer::AbstractVector) where {T}
+    elements_taken = 0 # number of elements taken so far
+    idx1 = firstindex(buffer)
+    target_buffer_len = min(n, c.sz_max)
+    lock(c)
+    try
+        while elements_taken < n && (isopen(c) || isready(c))
+            # wait until the channel has at least min_n elements or is full
+            while length(c.data) < target_buffer_len && isopen(c)
+                try
+                    wait(c.cond_take)
+                catch e
+                    if e isa InvalidStateException && e.state === :closed
+                        break
+                    else
+                        rethrow()
+                    end
+                end
+            end
+            # take as many elements as possible from the buffer
+            n_to_take = min(n - elements_taken, length(c.data))
+            idx_start = idx1 + elements_taken
+            idx_end = idx_start + n_to_take - 1
+            for (res_i, data_i) in Iterators.zip(idx_start:idx_end, eachindex(c.data))
+                @inbounds buffer[res_i] = c.data[data_i]
+            end
+            deleteat!(c.data, 1:n_to_take)
+            elements_taken += n_to_take
+            _increment_n_avail(c, -n_to_take)
+            foreach(_ -> notify(c.cond_put, nothing, true, false), 1:n_to_take)
```

Comment:
I personally think this looks nicer than a loop, but I don't know if the extra anon func is frowned upon